File: DsWorkJava/dswork.core/dswork/core/mybatis/dialect/OracleDialect.java
package dswork.core.mybatis.dialect;

/**
 * Dialect for Oracle
 * @author skey
 */
public class OracleDialect extends Dialect
{
	/**
	 * 是否支持分页，limit和offset
	 * @return boolean
	 */
	public boolean supportsLimitOffset()
	{
		return true;
	}
	
	/**
	 * 将sql变成分页sql
	 * @param sql 原始sql语句
	 * @param offset 跳过行数
	 * @param limit 返回行数
	 * @return String
	 */
	public String getLimitString(String sql, int offset, int limit)
	{
//		boolean hasUpdate = false;
//		if(sql.toLowerCase().endsWith(" for update"))
//		{
//			sql = sql.substring(0, sql.length() - 11);
//			hasUpdate = true;
//		}
		StringBuilder sb = new StringBuilder(sql.length() + 90);
//		if(offset > 0)
//		{
			sb.append("select * from ( select rownum rn, t_.* from (")
			.append(sql)
			.append(") t_ where rownum<=").append(offset + limit)
			.append(") n_ where n_.rn>").append(offset);
//		}
//		else
//		{
//			sb.append("select * from ( ");
//			sb.append(sql);
//			sb.append(" ) where rownum <= " + limit);
//		}
		//if(hasUpdate) {sb.append(" for update");};
		return sb.toString();
	}
}

File: DsWorkNet/Dswork.Core/Mybaits/Dialect/OracleDialect.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Dswork.Core.Mybaits.Dialect
{
	/// <summary>
	/// Dialect for Oracle
	/// </summary>
	public class OracleDialect : Dialect
	{
		/// <summary>
		/// 是否支持分页，limit和offset
		/// </summary>
		/// <returns>Boolean</returns>
		public override Boolean SupportsLimitOffset()
		{
			return true;
		}

		/// <summary>
		/// 将sql变成分页sql
		/// </summary>
		/// <param name="sql">原始sql语句</param>
		/// <param name="offset">跳过行数</param>
		/// <param name="limit">返回行数</param>
		/// <returns>String</returns>
		public override String GetLimitString(String sql, int offset, int limit)
		{
			StringBuilder sb = new StringBuilder(sql.Length + 90);
			sb.Append("select * from ( select t.*, rownum rn from ( ")
			.Append(sql)
			.Append(" ) t where rownum <= ").Append(offset + limit)
			.Append(" ) where rn > ").Append(offset);
			return sb.ToString();
		}
	}
}
